### compute classification accuracy using EMF RBM features

#### this could be automated

In [5]:
import numpy as np
import h5py

from sklearn.neural_network import BernoulliRBM
from sklearn import linear_model, datasets, metrics, preprocessing 
from sklearn.cross_validation import train_test_split
from sklearn.metrics import accuracy_score

### Read same MNIST data

In [6]:
hf =  h5py.File('mnist.h5','r')
print('List of arrays in this file: \n', hf.keys())
X = np.array(hf.get('HDF5.name___X'))
y = np.array(hf.get('HDF5.name___y'))

print X.shape, y.shape
hf.close()
print np.linalg.norm(X, ord=2)

('List of arrays in this file: \n', [u'HDF5.name___X', u'HDF5.name___y'])
(60000, 784) (60000,)
2117.63422548


In [7]:
hf =  h5py.File('mnist.tap3.fp.a10.h5','r')
print('List of arrays in this file: \n', hf.keys())
b = np.array(hf.get('mnist___bias'))
vb = np.array(hf.get('mnist___vbias'))
w = np.array(hf.get('mnist___weight'))

print b.shape, vb.shape, w.shape
hf.close()

('List of arrays in this file: \n', [u'mnist___bias', u'mnist___vbias', u'mnist___weight'])
(256,) (784,) (256, 784)


In [8]:
from sklearn.utils.fixes import expit    
from sklearn.utils.extmath import safe_sparse_dot

def sig_means(x, b, W):
    a = safe_sparse_dot(x, W.T) + b
    return expit(a, out=a)

In [9]:
p = sig_means(X, b, w)
X_train, X_test, Y_train, Y_test = train_test_split(p, y, test_size=0.2,random_state=0)
print X_train.shape, X_test.shape

for c in [5000]:
    lr  = linear_model.LogisticRegression()
    lr.C = c
    lr.fit(X_train, Y_train)
    Y_test_pred = lr.predict(X_test)
    acc = accuracy_score(Y_test, Y_test_pred)

    print c, acc

(48000, 256) (12000, 256)
5000 0.952666666667


mnist.tap2.fp.a3.h5  0.951  
mnist.tap2.fp.a6.h5  0.952916  
mnist.tap2.fp.a10.h5 0.9511666

mnist.tap3.fp.a3.h5  0.95125    
mnist.tap3.fp.a6.h5  0.95125 ?
mnist.tap3.fp.a10.h5 0.95267

mnist.cd.fp.a3.h5    0.95366  
mnist.cd.p.a3.h5     0.95325

mnist.cd.fp.a6.h5    0.95541    


In [ ]:
!ls *h5